iree/vm/stack.c

```diff
@@ -16,87 +16,706 @@
+// easier for us to add/version additional information or hide implementation
+// details.
+typedef struct iree_vm_stack_frame_header {
+  // Size, in bytes, of the frame header and frame payload including registers.
+  // Adding this value to the base header pointer will yield the next available
+  // memory location. Ensure that it does not exceed the total
+  // frame_storage_capacity.
+  iree_host_size_t frame_size;
+
+  // Pointer to the parent stack frame, usually immediately preceding this one
+  // in the frame storage. May be NULL.
+  struct iree_vm_stack_frame_header* parent;
+
+  // Stack frame type used to determine which fields are valid.
+  iree_vm_stack_frame_type_t type;
+
+  // Pointer to a register list within the stack frame where return registers
+  // will be stored by callees upon return.
+  //
+  // TODO(#1979): external frames will have a direct reference to the callee
+  // frame return registers.
+  const iree_vm_register_list_t* return_registers;
+
+  // Actual stack frame as visible through the API.
+  // The registers within the frame will (likely) point to addresses immediately
+  // following this header in memory.
+  iree_vm_stack_frame_t frame;
+} iree_vm_stack_frame_header_t;
+
+// Core stack storage. This will be mapped either into dynamic memory allocated
+// by the member allocator or static memory allocated externally. Static stacks
+// cannot grow when storage runs out while dynamic ones will resize their stack.
+struct iree_vm_stack {
+  // NOTE: to get better cache hit rates we put the most frequently accessed
+  // members first.
+
+  // Pointer to the current top of the stack.
+  // This can be used to walk the stack from top to bottom by following the
+  // |parent| pointers. Note that these pointers are invalidated each time the
+  // stack grows (if dynamic growth is enabled) and all of the frames will need
+  // updating.
+  iree_vm_stack_frame_header_t* top;
+
+  // Base pointer to stack storage.
+  // For statically-allocated stacks this will (likely) point to immediately
+  // after the iree_vm_stack_t in memory. For dynamically-allocated stacks this
+  // will (likely) point to heap memory.
+  iree_host_size_t frame_storage_capacity;
+  iree_host_size_t frame_storage_size;
+  void* frame_storage;
+
+  // True if the stack owns the frame_storage and should free it when it is no
+  // longer required. Host stack-allocated stacks don't own their storage but
+  // may transition to owning it on dynamic growth.
+  bool owns_frame_storage;
+
+  // Resolves a module to a module state within a context.
+  // This will be called on function entry whenever module transitions occur.
+  iree_vm_state_resolver_t state_resolver;
+
+  // Allocator used for dynamic stack allocations. May be the null allocator
+  // if growth is prohibited.
+  iree_allocator_t allocator;
+};
+
+//===----------------------------------------------------------------------===//
+// Math utilities, kept here to limit dependencies
```

Comment:
Want to add `inline` to any of these? (assuming we can in the version(s) of C that this is compiled with)